### ISSUE TITLE
`active_curve` index out of range when graphing complicated functions

### ISSUE THREAD

**Issue by aheze** [jared-hughes/isosurfaces#5]
Hi! I came across your library and it works really well. But when I tried graphing the expression $`\tan(x^2+y^2) = 1`$, represented using this function:

```py
def f(x, y):
    return np.tan(x ** 2 + y ** 2) - 1
```
... I got this:

```
Traceback (most recent call last):
  File "/Users/aheze/Downloads/isosurfaces-main/examples/isoline_demo.py", line 23, in <module>
    curves = CurveTracer(triangles, fn, tol).trace()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/isosurfaces/isoline.py", line 234, in trace
    self.march_triangle(triangle)
  File "/opt/homebrew/lib/python3.11/site-packages/isosurfaces/isoline.py", line 256, in march_triangle
    self.active_curve.append(self.active_curve[0])
                             ~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

---

So I added a check to see if `self.active_curve` is empty inside [`march_triangle`](https://github.com/jared-hughes/isosurfaces/blob/main/isosurfaces/isoline.py#L242):

```py
if closed_loop:
    
    # check if it's not empty
    if len(self.active_curve) > 0:
    
        # close back the loop
        self.active_curve.append(self.active_curve[0])
```

This time, it ran without crashing. Here's the result:

![Multiple circles nested in each other with line fragments in between](https://github.com/jared-hughes/isosurfaces/assets/49819455/694e14c1-736e-4bc9-b1b2-ed244bf23983)

There's a bunch of weird line fragments, probably because `active_curve` isn't being closed off. For reference here is the expected result from Desmos:

![Expected result with multiple circles nested in each other](https://github.com/jared-hughes/isosurfaces/assets/49819455/bb6062b0-56c1-4c3a-a111-1535dafaca5b)

Do you have any idea what's going on? Again, thanks for your work on this library, it's exactly what I've been looking for!

**Comment by jared-hughes** [comment]
I suspect closing off the loop is the wrong behavior in some cases. You might get better results from simply deleting the line
```py
        self.active_curve.append(self.active_curve[0])
```
On the other hand, closing off the loop is necessary for a circle like `x^2+y^2=1`. Otherwise it would just be missing the segment to complete the circle.

**Comment by jared-hughes** [comment]
Figured out the problem. I falsely assumed that a point would be added to the curve if `triangle.next` is not None. But there's a discrepancy:
- `triangle.next` is not None when the vertices have opposite sign
- A point is added when the vertices have opposite sign *and* binary search finds a zero between them

Binary search finds a zero in all continuous cases, but tan() has an asymptote, so it can end up finding two points with evaluation tending to -Infinity and +Infinity. So it rejects adding that point.

I've fixed this in https://github.com/jared-hughes/isosurfaces/commit/92372b7726bdf0f79a501b248e877516398512d5
